Replace sprintf() with std::string and/or format().  Does part of the…

backwardsEric opened the pull request:
… work of resolving https://github.com/hengband/hengband/issues/2858 and https://github.com/hengband/hengband/issues/2859 .

This is a subset of the changes in #2861 as requested by feedback there.  It does include some simpler replacements for sprintf() (strnfmt(), angband_strcpy(), or avoiding copying altogether) where those replacements are used in the same file as replacements that use std::string/format().  In general, these changes don't change the return value or declared arguments for functions.  There's two exceptions to that in wizard/artifact-analyzer.cpp where two local functions pick up an additional size_t argument to specify the length of a buffer.

habu1010 commented on src/wizard/monster-info-spoiler.cpp:
```diff
@@ -203,32 +200,22 @@ SpoilerOutputResultType spoil_mon_info(concptr fname)
+        spoil_out(std::string(format("Num:%d  ", enum2i(r_idx))).data());
+        spoil_out(std::string(format("Lev:%d  ", (int)r_ptr->level)).data());
+        spoil_out(std::string(format("Rar:%d  ", r_ptr->rarity)).data());
+        spoil_out(std::string(format("Spd:+d  ", r_ptr->speed - STANDARD_SPEED)).data());
```
`%+d` has become `+d`.

habu1010 commented on src/wizard/monster-info-spoiler.cpp:
```diff
@@ -203,32 +200,22 @@ SpoilerOutputResultType spoil_mon_info(concptr fname)
-        spoil_out(buf);
-        sprintf(buf, "Spd:%+d  ", r_ptr->speed - STANDARD_SPEED);
-        spoil_out(buf);
+        spoil_out(std::string(format(" '%c'\n)", r_ptr->d_char)).data());
```
")\n" has become "\n)".

habu1010 commented on src/wizard/wizard-spoiler.cpp:
```diff
@@ -212,23 +210,19 @@ static SpoilerOutputResultType spoil_player_spell(concptr fname)
         for (int16_t r = 1; r < MAX_MAGIC; r++) {
-            sprintf(buf, "[Realm: %s]\n", realm_names[r]);
-            spoil_out(buf);
+            spoil_out(std::string("[Realm: ").append(realm_names[r]).append("\n").data());
```
"]\n" has become "\n".

backwardsEric commented:
The forced push should have handled the places you noted.  Other changes included in that are:

1. Since the changes to spoil_player_spell() already had it calling format(), got rid of the uses of append() there and made use of format() instead.
2. In view/display-self-info.cpp changed vir_name to a std::string_view and in status/shape-changer.cpp changed sex_msg to a std::string_view since both were simply referring to other string constants or character arrays and didn't need to make a copy.
3. In market/arena.cpp, rearranged how the name is set up so that it is hopefully more readable.
4. Changed more instances of fputs() to fprintf() in knowledge/knowledge-quests.cpp to avoid the extra step of setting up the string passed to fputs().
5. Changed cmd-io/cmd-floor.cpp to grab the direction string from a lookup table rather than building it by concatenation.  Slightly changed the English version of that string to be more idiomatic (no capitalization and include "of" if the sector isn't the current one).

habu1010 commented:
After all, the current format function's specification of returning a pointer to a static internal buffer must be used with great care, so I am currently trying to change the return value of the format function to a std::string.

Since the current specification makes it burdensome to determine whether the format function can be used or not, I think it would be better to rebase this pull request and the rest of the pull requests after the format function return value is fixed to be a std::string. What do you think?

backwardsEric commented:
There's a subset of the changes here that aren't using format() or can be fairly easily proved to not be any riskier than the current code (the changed function or the file-level static functions it uses were already calling format() in the current code and immediately using its value).  I can split those off into a separate pull request if you deem it worthwhile.

habu1010 commented:
I was asking your opinion if there is any place where it would be possible to replace sprintf with clearer code than it is now, once the format function returns std::string and is guaranteed to be safe to use at all times.

For example, I believe #2953 could be rewritten using format while keeping the original sprintf formatting.

If there is nothing in this pull request that could be rewritten using the safe format function, then it is fine as is.

backwardsEric commented:
A safe format() with no buffer length limitation would change a lot of the (probably poor) decisions that went into the sprintf() replacements.  It would change #2953 , but there are other issues with that one.  While we're changing that file, do we want to address the other problems there:

1. be able to handle of terms whose width or height are greater than 255 (or 127 if char is signed and there's no provision to wrap negative coordinates when reading back the movie)
2. in the text hook, there's provisions for keeping the length of the 'n' command less than 128 (presumably to avoid a negative length if chars are signed) but no similar provision to that for the 't' command; breaking up the buffer for the 't' command has the complication of trying to avoid splitting up a multi-byte encoding of a kanji character
3. (minor) extend the run-length compression for repeated bytes in the text hook so it can be used it for long enough repeated sequences that don't happen to be the entire buffer passed to the hook

backwardsEric commented:
The latest forced push made these changes:

1. `if (rand_tmp_str.length())' -> 'if (!rand_tmp_str.empty())' at knowledge/knowledge-quests.cpp:177
2. remove `>` inserted into Japanese text at market/play-gamble.cpp:163.
3. since string literal was the same in both the if and else branches at status/shape-changer.cpp:151, moved the common code out of the if/else

habu1010 commented on src/spell/spell-info.cpp:
```diff
@@ -296,25 +295,25 @@ void print_spells(PlayerType *player_ptr, SPELL_IDX target_spell, SPELL_IDX *spe
-        strcpy(info, spell_info->data());
-        concptr comment = info;
+        std::string info = exe_spell(player_ptr, use_realm, spell, SpellProcessType::INFO)->data();
+        concptr comment = info.data();
```
exe_spell returns type `std::optional<std::string>`.
The '->' operator of std::optional calls a member of the type it holds (in this case, std::string).
Thus, the following code

```
std::string info = exe_spell(player_ptr, use_realm, spell, SpellProcessType::INFO)->data();
```

calls data() on the returned string type object, and then creates the std::string type again from the `const char*`, which is a two-step process.

To get the string type directly (which should not return std::nullopt if called with SpellProcessType::INFO), use

```
auto info = exe_spell(player_ptr, use_realm, spell, SpellProcessType::INFO).value();
```
is appropriate.

backwardsEric replied:
That line is now back to what is in the current code, `const auto info = ...`, and the initialization of comment just after it has changed.

habu1010 commented on src/view/display-self-info.cpp:
```diff
@@ -24,9 +26,7 @@ void display_max_base_status(PlayerType *player_ptr, self_info_type *self_ptr)
-        char stat_desc[80];
-        sprintf(stat_desc, "%s 18/%d", stat_names[v_nr], player_ptr->stat_max_max[v_nr] - 18);
-        strcpy(self_ptr->s_string[v_nr], stat_desc);
+        angband_strcpy(self_ptr->s_string[v_nr], format("%s 18/%d", stat_names[v_nr], player_ptr->stat_max_max[v_nr] - 18), sizeof(self_ptr->s_string[v_nr]));
```
Could it be rewritten using strnfmt?

backwardsEric replied:
Done in the latest forced push.

habu1010 commented on src/view/display-self-info.cpp:
```diff
@@ -35,44 +35,42 @@ void display_virtue(PlayerType *player_ptr, self_info_type *self_ptr)
     for (int v_nr = 0; v_nr < 8; v_nr++) {
-        GAME_TEXT vir_name[20];
-        char vir_desc[80];
+        std::string_view vir_name = virtue[(player_ptr->vir_types[v_nr]) - 1];
```
In this case, it is better to use auto for vir_name (resulting in concptr type).
If we use std::string_view, we need to add data() after each view, and it looks too noisy.

habu1010 commented on src/wizard/monster-info-spoiler.cpp:
```diff
@@ -203,32 +200,22 @@ SpoilerOutputResultType spoil_mon_info(concptr fname)
 
-        sprintf(buf, _("%s/%s  (", "%s%s ("), r_ptr->name.data(), _(r_ptr->E_name.data(), "")); /* ---)--- */
-        spoil_out(buf);
+        spoil_out(std::string(format(_("%s/%s  (", "%s%s ("), r_ptr->name.data(), _(r_ptr->E_name.data(), ""))).data()); /* ---)--- */
```
Since the process of calling fromat is not done in spoil_out, shouldn't it be ok to not wrap it in std::string().data()?

(I may have to add .data() when I make the format function return a std::string, but even then std::string() would not be needed.)

backwardsEric replied:
Since none of the existing uses of spoil_out() were directly passed the result of format(), I wrapped format() with std::string() as a way of protecting it against some future change that would have spoil_out() call format() internally.  With your planned change to have format() return a std::string, I can remove that wrapping.

habu1010 commented on src/wizard/wizard-spoiler.cpp:
```diff
@@ -188,17 +188,15 @@ static SpoilerOutputResultType spoil_player_spell(concptr fname)
     put_version(title);
-    sprintf(buf, "Player Spells for %s\n", title);
-    spoil_out(buf);
+    spoil_out(std::string(format("Player spells for %s\n", title)).data());
```
Likewise above.

habu1010 reviewed (commented):
Some comments added.
Once these have been addressed, I think they can be merged, but are there any other areas you would like to make additional corrections?

backwardsEric commented:
Besides the things you noted, there was one correction I made (remove an extra `)` that crept into the string at wizard/wizard/monster-info-spoiler.cpp:205.  Other than that, I don't have any more corrections at this point.

habu1010 commented:
Understood.
I think there are no more outstanding problems for now, so I will approve and merge.
Thank you very much for your long term support.
